### PR TITLE
Add HTTP2 to gRPC server state machine

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

We use the HTTP 2-to-1 codecs on the server to simplify the
implementation. However this comes at a performance cost. We removed
the HTTP/1 types from surface API in #1017; this change introduces a
state machine for a translating between HTTP2 frame payloads and gRPC
server request parts. Future PRs will build on top of this to replace
both the 'HTTP1ToGRPCServerCodec' and the 'GRPCServerRequestRoutingHandler'.

Modifications:

Add an 'HTTP2ToRawGRPCStateMachine' and tests.

Result:

We have a state machine in place on which we can build a channel
handler.

### CODE REVIEW

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift [hunk 1]
```diff
@@ -0,0 +1,1277 @@
...
+  }
+
+  /// Extract and validate the request message encoding header.
+  /// - Parameters:
+  ///   - headers: The headers to extract the message encoding header from.
+  /// - Returns: `RequestEncodingValidation`, either a message reader suitable for decoding requests
+  ///   and an accept encoding response header if the request encoding was valid, or a pair of
+  ///     `GRPCStatus` and trailers to close the RPC with.
+  private func extractRequestEncoding(from headers: HPACKHeaders) -> RequestEncodingValidation {
+    let encodingHeader = headers.first(name: GRPCHeaderName.encoding)
```
Comment (Lukasa) [inline Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:460]: Is the use of `first` safe here? Are we choosing to ignore other appearances of this header on purpose?

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift [hunk 2]
```diff
@@ -0,0 +1,1277 @@
...
+    contentType: ContentType,
+    responseEncoding: String?,
+    acceptableRequestEncoding: String?,
+    userProvidedHeaders: HPACKHeaders,
+    normalizeUserProvidedHeaders: Bool
+  ) -> HPACKHeaders {
+    let capacity = 2 // :status, content-type
+      + (responseEncoding == nil ? 0 : 1)
+      + (acceptableRequestEncoding == nil ? 0 : 1)
+      + userProvidedHeaders.count
```
Comment (Lukasa) [inline Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:1182]: Is the math worthwhile here, or should we just do 4 + count? Seems like excessive branching for minimal savings.

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift [hunk 3]
```diff
@@ -0,0 +1,1277 @@
...
+    for status: GRPCStatus,
+    contentType: ContentType,
+    acceptableRequestEncoding: String?,
+    userProvidedHeaders: HPACKHeaders?,
+    normalizeUserProvidedHeaders: Bool
+  ) -> HPACKHeaders {
+    let capacity = 3 // :status, status-code, content-type
+      + (status.message == nil ? 0 : 1)
+      + (acceptableRequestEncoding == nil ? 0 : 1)
+      + (userProvidedHeaders.map { $0.count } ?? 0)
```
Comment (Lukasa) [inline Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:1214]: Same question here about the math

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift [hunk 4]
```diff
@@ -0,0 +1,1277 @@
...
+    return headers
+  }
+
+  static func makeResponseTrailers(
+    for status: GRPCStatus,
+    userProvidedHeaders: HPACKHeaders,
+    normalizeUserProvidedHeaders: Bool
+  ) -> HPACKHeaders {
+    let capacity = 1 // status-code
+      + (status.message == nil ? 0 : 1)
```
Comment (Lukasa) [inline Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:1246]: Same question here about math.

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift (hunk 1 above)
Comment (glbrntt) [reply Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:460]: Good question. We don't expect other occurrences of this header which is why I used `first`. Do you think we should use `canonicalForm` instead and close if there's more than one value?

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift (hunk 2 above)
Comment (glbrntt) [reply Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:1182]: You're probably right tbh.

File: Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift (hunk 1 above)
Comment (Lukasa) [reply Sources/GRPC/HTTP2ToRawGRPCStateMachine.swift:460]: I'd be inclined to do that, yeah.

**Review by Lukasa — APPROVED** [review]

**Pull request merged**